### PR TITLE
feat(core): add support for multi use invitations

### PR DESCRIPTION
**Pull request by TimoGlastra**
Signed-off-by: Timo Glastra <timo@animo.id>

### CODE REVIEW

**Comment by TimoGlastra** [comment]
Could someone review please?

File: packages/core/src/modules/connections/services/ConnectionService.ts [hunk 1]
```diff
@@ -194,6 +199,25 @@ export class ConnectionService {
       throw new AriesFrameworkError('Invalid message')
     }
 
+    // Create new connection if using a multi use invitation
+    if (connectionRecord.multiUseInvitation) {
```
Comment (blu3beri) [inline packages/core/src/modules/connections/services/ConnectionService.ts:203]: Could be combined with the if statement below.

**Review by blu3beri — APPROVED** [review]
Not too knowledgeale with the multi use specifications, but code-wise it looks good!

**Comment by genaris** [comment]
Nice and long-awaited feature!

File: packages/core/tests/connections.test.ts [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    faberAliceConnection1 = await faberAgent.connections.returnWhenIsConnected(faberAliceConnection1.id)
+    faberAliceConnection2 = await faberAgent.connections.returnWhenIsConnected(faberAliceConnection2.id)
+
+    expect(faberAliceConnection1).toBeConnectedWith(aliceFaberConnection1)
+    expect(faberAliceConnection2).toBeConnectedWith(aliceFaberConnection2)
+
+    const faberConnection = await faberAgent.connections.getById(faberConnectionId)
+    // Expect initial connection to still be in state invited
+    return expect(faberConnection.state).toBe(ConnectionState.Invited)
+  })
```
[the remark this answers is not in the document]
Comment (TimoGlastra) [reply packages/core/tests/connections.test.ts:85]: Yes! The first record will always stay in state `Invited` and we create new records when people send requests for the invitation. This is how both ACA-Py and .NET have also implemented it

File: packages/core/src/modules/connections/handlers/ConnectionRequestHandler.ts [hunk 3]
```diff
@@ -9,23 +10,38 @@ import { ConnectionRequestMessage } from '../messages'
...
 
-    if (messageContext.connection?.autoAcceptConnection ?? this.agentConfig.autoAcceptConnections) {
-      const { message } = await this.connectionService.createResponse(messageContext.connection.id)
-      return createOutboundMessage(messageContext.connection, message)
+    // routing object is required for multi use invitation, because we're creating a
+    // new keypair that possibly needs to be registered at a mediator
+    if (messageContext.connection.multiUseInvitation) {
+      const mediationRecord = await this.mediationRecipientService.discoverMediation()
+      routing = await this.mediationRecipientService.getRouting(mediationRecord)
+    }
```
[the remark this answers is not in the document]
Comment (TimoGlastra) [reply packages/core/src/modules/connections/handlers/ConnectionRequestHandler.ts:38]: Maybe not. But I would rather take all edge cases into account for these things. Otherwise we need good documentation on what features do and don't work together :)

File: packages/core/tests/connections.test.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (jakubkoci) [reply packages/core/tests/connections.test.ts:85]: Interesting 🤔
